## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [2]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [3]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [4]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [30]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df["job"].unique()
df["order_year"].unique()
df["client_id"].unique()
df["category"].unique()
df["subcategory"].unique()
df['item_id'].unique()




array(['EUD29711-63-6U', 'XDA18116-89-4A', 'ABE59463-05-7E',
       'ZMM00836-65-0C', 'BZX55559-12-3X', 'BMB23158-88-9A',
       'UAC50632-25-4E', 'UEU25316-01-1U', 'DDA06248-98-7Z',
       'DUU85042-87-5E', 'DCD61783-36-6B', 'ZAE56533-39-8U',
       'BDC65078-60-7M', 'AME10969-48-7D', 'BXM59071-01-6D',
       'CCX95657-71-3B', 'ZEA01443-18-5A', 'CUU90642-03-7C',
       'UCX86104-19-0Z', 'ZDX29223-32-0B', 'ZAU36048-03-9A',
       'EZX36378-17-7A', 'UAU40137-77-3U', 'ZMB00381-41-4E',
       'XZZ76042-95-6X', 'BAD54487-56-3C', 'UZZ84388-36-3B',
       'DBE42716-07-1U', 'UMC49136-12-1E', 'EDU54097-48-4E',
       'MAU13971-98-9B', 'ZEZ40347-39-4X', 'UUE15482-56-0X',
       'CCA32097-88-6M', 'AEE51706-20-3U', 'DZX13514-89-1M',
       'CZC97008-29-8E', 'AEM31850-24-5M', 'MDB11336-71-2E',
       'XAU13501-34-9A', 'CBZ89457-17-5C', 'AED36322-86-8U',
       'AZB92583-19-5D', 'DXU77211-79-0B', 'EUX51497-34-6B',
       'AEE37351-26-7B', 'XBM94536-93-1U', 'AUZ65769-43-0D',
       'EBB79610-74-2A',

In [6]:
# What three item categories had the most entries?
df["category"].value_counts().head(3)


category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [7]:
# For the category with the most entries, which subcategory had the most entries?
df["subcategory"].value_counts().head(1)



subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [8]:
# Which five clients had the most entries in the data?
df["client_id"].value_counts().head()

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [9]:
# Store the client ids of those top 5 clients in a list.
top_5_client_id = [33615, 66037, 46820, 38378, 24741]
top_5_client_id

[33615, 66037, 46820, 38378, 24741]

In [56]:
# How many total units (the qty column) did the client with the most entries order order?
filtered_33615_df = df[df["client_id"] == 33615]
total_qty_33615_df = filtered_33615_df["qty"].sum()
total_qty_33615 = total_qty_33615_df
print(f'Client ID 33615 had the most entries. They ordered a total of {total_qty_33615: ,.2f} units.')


Client ID 33615 had the most entries. They ordered a total of  64,313.00 units.


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [12]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df["line_subtotal"] = df["unit_price"] * df["qty"]
df[["unit_price", "qty", "line_subtotal"]].head()


unit_price  qty  line_subtotal
0     1096.80  105      115164.00
1       24.95   21         523.95
2       13.52   39         527.28
3       36.42   29        1056.18
4      195.10   20        3902.00

In [13]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
df["total_weight"] = df["unit_weight"] * df["qty"]
df[["unit_weight", "qty", "total_weight"]]
def shipping_price(total_weight):
    return total_weight * 7 if total_weight > 50 else total_weight * 10
df["shipping_price"] = df["total_weight"].apply(shipping_price)
df[["unit_price","unit_weight", "qty", "total_weight", "shipping_price"]].head()

    

unit_price  unit_weight  qty  total_weight  shipping_price
0     1096.80         7.50  105        787.50         5512.50
1       24.95         1.49   21         31.29          312.90
2       13.52         1.68   39         65.52          458.64
3       36.42         1.23   29         35.67          356.70
4      195.10        46.43   20        928.60         6500.20

In [14]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df["sub_total"] = df["shipping_price"] + df["line_subtotal"]
df["line_price"] = round(df["sub_total"] * 1.0925, 2)
df[["line_subtotal", "shipping_price", "sub_total", "line_price"]].head()


line_subtotal  shipping_price  sub_total  line_price
0      115164.00         5512.50  120676.50   131839.08
1         523.95          312.90     836.85      914.26
2         527.28          458.64     985.92     1077.12
3        1056.18          356.70    1412.88     1543.57
4        3902.00         6500.20   10402.20    11364.40

In [15]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df["line_cost"] = df["unit_cost"] * df["qty"] + df["shipping_price"]
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   
3          17        2023  ...     24.85        1.23   29            3   
4          21        2023  ...    108.17       46.43   20            1   

   line_subtotal  total_weight  shipping_price  sub_total  line_price  \
0      115164.00        787.50         5512.50  120676.50   131839.08   
1         523.95         31.29          312.90     836.85      914.26   
2         527.28         65.52          458.64     985.92     1077.12   
3        1056.18         35.67          356.70    1412.88     1543.57   
4        3902.00        928.60         6500.20   10402.20    11364.40   

   line_cost  
0   85597.05  
1     629.79  
2     765.18  
3    1077.35  
4    8663.60  

[5 rows x 24 columns]

In [16]:
# Create a column for the profit of each line using line cost and line price
df["line_profit"] = df["line_price"] - df["line_cost"]
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_weight  qty line_number  line_subtotal  \
0          17        2023  ...        7.50  105           1      115164.00   
1          20        2023  ...        1.49   21           0         523.95   
2           4        2023  ...        1.68   39           6         527.28   
3          17        2023  ...        1.23   29           3        1056.18   
4          21        2023  ...       46.43   20           1        3902.00   

   total_weight  shipping_price  sub_total  line_price  line_cost  line_profit  
0        787.50         5512.50  120676.50   131839.08   85597.05     46242.03  
1         31.29          312.90     836.85      914.26     629.79       284.47  
2         65.52          458.64     985.92     1077.12     765.18       311.94  
3         35.67          356.70    1412.88     1543.57    1077.35       466.22  
4        928.60         6500.20   10402.20    11364.40    8663.60      2700.80  

[5 rows x 25 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [23]:
# Check your work using the totals above
order_2742071_df = df.loc[df["order_id"] == 2742071]
total_2742071_df = round(order_2742071_df["line_price"].sum(), 2)
total_2742071 = total_2742071_df
print(f'Order 2742071 had a total price of ${total_2742071: ,.2f}.')

order_2173913_df = df.loc[df["order_id"] == 2173913]
total_2173913_df = round(order_2173913_df["line_price"].sum(), 2)
total_2173913 = total_2173913_df
print(f'Order 2173913 had a total price of ${total_2173913: ,.2f}.')

order_6128929_df = df.loc[df["order_id"] == 6128929]
total_6128929_df = round(order_6128929_df["line_price"].sum(), 2)
total_6128929 = total_6128929_df
print(f'Order 6128929 had a total price of ${total_6128929: ,.2f}.')

Order 2742071 had a total price of $ 152,811.89.
Order 2173913 had a total price of $ 162,388.71.
Order 6128929 had a total price of $ 923,441.25.


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [24]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
client_33615_df = df.loc[df["client_id"] == 33615]
total_client_33615_df = round(client_33615_df["line_price"].sum(), 2)
total_client_33615 = total_client_33615_df
print(f'Client ID 33615 spent a total amount of ${total_client_33615: ,.2f}.')

client_66037_df = df.loc[df["client_id"] == 66037]
total_client_66037_df = round(client_66037_df["line_price"].sum(), 2)
total_client_66037 = total_client_66037_df
print(f'Client ID 66037 spent a total amount of ${total_client_66037: ,.2f}.')

client_46820_df = df.loc[df["client_id"] == 46820]
total_client_46820_df = round(client_46820_df["line_price"].sum(), 2)
total_client_46820 = total_client_46820_df
print(f'Client ID 46820 spent a total amount of ${total_client_46820: ,.2f}.')

client_38378_df = df.loc[df["client_id"] == 38378]
total_client_38378_df = round(client_38378_df["line_price"].sum(), 2)
total_client_38378 = total_client_38378_df
print(f'Client ID 38378 spent a total amount of ${total_client_38378: ,.2f}.')

client_24741_df = df.loc[df["client_id"] == 24741]
total_client_24741_df = round(client_24741_df["line_price"].sum(), 2)
total_client_24741 = total_client_24741_df
print(f'Client ID 24741 spent a total amount of ${total_client_24741: ,.2f}.')

Client ID 33615 spent a total amount of $ 8,377,308.52.
Client ID 66037 spent a total amount of $ 10,259,514.79.
Client ID 46820 spent a total amount of $ 9,743,794.36.
Client ID 38378 spent a total amount of $ 12,906,550.87.
Client ID 24741 spent a total amount of $ 82,268,892.02.


In [25]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
client_33615_df = df.loc[df["client_id"] == 33615]
total_units_purchased_33615_df = round(client_33615_df["qty"].sum(), 2)
total_units_purchased_33615_df
total_shipping_price_33615_df = round(client_33615_df["shipping_price"].sum(), 2)
total_shipping_price_33615_df
total_line_price_33615_df = round(client_33615_df["line_price"].sum(), 2)
total_line_price_33615_df
total_line_cost_33615_df = round(client_33615_df["line_cost"].sum(), 2)
total_line_cost_33615_df
total_line_profit_33615_df = round(client_33615_df["line_profit"].sum(), 2)
total_line_profit_33615_df


client_66037_df = df.loc[df["client_id"] == 66037]
total_units_purchased_66037_df = round(client_66037_df["qty"].sum(), 2)
total_units_purchased_66037_df
total_shipping_price_66037_df = round(client_66037_df["shipping_price"].sum(), 2)
total_shipping_price_66037_df
total_line_price_66037_df = round(client_66037_df["line_price"].sum(), 2)
total_line_price_66037_df
total_line_cost_66037_df = round(client_66037_df["line_cost"].sum(), 2)
total_line_cost_66037_df
total_line_profit_66037_df = round(client_66037_df["line_profit"].sum(), 2)
total_line_profit_66037_df


client_46820_df = df.loc[df["client_id"] == 46820]
total_units_purchased_46820_df = round(client_46820_df["qty"].sum(), 2)
total_units_purchased_46820_df
total_shipping_price_46820_df = round(client_46820_df["shipping_price"].sum(), 2)
total_shipping_price_46820_df
total_line_price_46820_df = round(client_46820_df["line_price"].sum(), 2)
total_line_price_46820_df
total_line_cost_46820_df = round(client_46820_df["line_cost"].sum(), 2)
total_line_cost_46820_df
total_line_profit_46820_df = round(client_46820_df["line_profit"].sum(), 2)
total_line_profit_46820_df


client_38378_df = df.loc[df["client_id"] == 38378]
total_units_purchased_38378_df = round(client_38378_df["qty"].sum(), 2)
total_units_purchased_38378_df
total_shipping_price_38378_df = round(client_38378_df["shipping_price"].sum(), 2)
total_shipping_price_38378_df
total_line_price_38378_df = round(client_38378_df["line_price"].sum(), 2)
total_line_price_38378_df
total_line_cost_38378_df = round(client_38378_df["line_cost"].sum(), 2)
total_line_cost_38378_df
total_line_profit_38378_df = round(client_38378_df["line_profit"].sum(), 2)
total_line_profit_38378_df


client_24741_df = df.loc[df["client_id"] == 24741]
total_units_purchased_24741_df = round(client_24741_df["qty"].sum(), 2)
total_units_purchased_24741_df
total_shipping_price_24741_df = round(client_24741_df["shipping_price"].sum(), 2)
total_shipping_price_24741_df
total_line_price_24741_df = round(client_24741_df["line_price"].sum(), 2)
total_line_price_24741_df
total_line_cost_24741_df = round(client_24741_df["line_cost"].sum(), 2)
total_line_cost_24741_df
total_line_profit_24741_df = round(client_24741_df["line_profit"].sum(), 2)
total_line_profit_24741_df


summary_top_5_df = pd.DataFrame({
    'client_id': [33615, 66037, 46820, 38378, 24741],
    'total_units_purchased': [total_units_purchased_33615_df, total_units_purchased_66037_df, total_units_purchased_46820_df, total_units_purchased_38378_df, total_units_purchased_24741_df],
    'total_shipping_price': [total_shipping_price_33615_df, total_shipping_price_66037_df, total_shipping_price_46820_df, total_shipping_price_38378_df, total_shipping_price_24741_df],
    'total_line_price': [total_line_price_33615_df, total_line_price_66037_df, total_line_price_46820_df, total_line_price_38378_df, total_line_price_24741_df],
    'total_line_cost':[total_line_cost_33615_df, total_line_cost_66037_df, total_line_cost_46820_df, total_line_cost_38378_df, total_line_cost_24741_df],
    'total_line_profit': [total_line_profit_33615_df, total_line_profit_66037_df, total_line_profit_46820_df, total_line_profit_38378_df, total_line_profit_24741_df]
})

summary_top_5_df


client_id  total_units_purchased  total_shipping_price  total_line_price  \
0      33615                  64313            1828984.89        8377308.52   
1      66037                  43018            1395151.85       10259514.79   
2      46820                  75768            1601448.84        9743794.36   
3      38378                  73667            3429455.40       12906550.87   
4      24741                 239862            5126448.37       82268892.02   

   total_line_cost  total_line_profit  
0       6175313.91         2201994.61  
1       7004482.98         3255031.81  
2       7007191.64         2736602.72  
3       9634720.98         3271829.89  
4      45688899.71        36579992.31

In [26]:
# Format the data and rename the columns to names suitable for presentation.
rename_top_5_df = summary_top_5_df.rename(columns={"client_id": "Client ID", "total_units_purchased": "Units", "total_shipping_price": "Shipping", "total_line_price": "Total Revenue", "total_line_cost": "Total Cost", "total_line_profit": "Total Profit"})

# Define the money columns. 
money_columns = ["Shipping", "Total Revenue", "Total Cost", "Total Profit"]

# Define a function that converts a dollar amount to millions.
def currency_format_millions(dollar_amount):
    return dollar_amount / 1000000
    
# Apply the currency_format_millions function to only the money columns. 
rename_top_5_df[money_columns] = rename_top_5_df[money_columns].apply(currency_format_millions, axis = 1)

# Rename the columns to reflect the change in the money format. 
rename_df = rename_top_5_df.rename(columns={"Shipping": "Shipping (millions)", "Total Revenue": "Total Revenue (millions)", "Total Cost": "Total Cost (millions)", "Total Profit": "Total Profit (millions)"})

rename_df[["Client ID", "Units", "Shipping (millions)", "Total Revenue (millions)", "Total Cost (millions)", "Total Profit (millions)"]].head()

Client ID   Units  Shipping (millions)  Total Revenue (millions)  \
0      33615   64313             1.828985                  8.377309   
1      66037   43018             1.395152                 10.259515   
2      46820   75768             1.601449                  9.743794   
3      38378   73667             3.429455                 12.906551   
4      24741  239862             5.126448                 82.268892   

   Total Cost (millions)  Total Profit (millions)  
0               6.175314                 2.201995  
1               7.004483                 3.255032  
2               7.007192                 2.736603  
3               9.634721                 3.271830  
4              45.688900                36.579992

In [27]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
updated_df = rename_df.sort_values(by='Total Profit (millions)', ascending=False)
updated_df

Client ID   Units  Shipping (millions)  Total Revenue (millions)  \
4      24741  239862             5.126448                 82.268892   
3      38378   73667             3.429455                 12.906551   
1      66037   43018             1.395152                 10.259515   
2      46820   75768             1.601449                  9.743794   
0      33615   64313             1.828985                  8.377309   

   Total Cost (millions)  Total Profit (millions)  
4              45.688900                36.579992  
3               9.634721                 3.271830  
1               7.004483                 3.255032  
2               7.007192                 2.736603  
0               6.175314                 2.201995

In [49]:
#What is Client 33615's purchase history? 
client_id_33615 = 33615
purchase_history_33615_df = df[df['client_id'] == client_id_33615]
purchase_history_33615_df[['order_id', 'category', 'subcategory']]
purchase_history_33615_df['category'].value_counts()


category
consumables    89
furniture      52
software       36
decor          31
electronics    12
Name: count, dtype: int64

In [50]:
#What is Client 46820's purchase history? 
client_id_46820 = 46820
purchase_history_46820_df = df[df['client_id'] == client_id_46820]
purchase_history_46820_df[['order_id', 'category', 'subcategory']]
purchase_history_46820_df['category'].value_counts()

category
consumables    78
decor          42
furniture      41
software       38
electronics    10
Name: count, dtype: int64

In [53]:
#What is Client 66037's purchase history? 
client_id_66037 = 66037
purchase_history_66037_df = df[df['client_id'] == client_id_66037]
purchase_history_66037_df[['order_id', 'category', 'subcategory']]
purchase_history_66037_df['category'].value_counts()

category
consumables    95
furniture      38
software       34
decor          31
electronics    13
Name: count, dtype: int64

In [54]:
#What is Client 38378's purchase history? 
client_id_38378 = 38378
purchase_history_38378_df = df[df['client_id'] == client_id_38378]
purchase_history_38378_df[['order_id', 'category', 'subcategory']]
purchase_history_38378_df['category'].value_counts()

category
consumables    90
software       46
furniture      33
decor          28
electronics    10
Name: count, dtype: int64

In [55]:
#What is Client 24741's purchase history? 
client_id_24741 = 24741
purchase_history_24741_df = df[df['client_id'] == client_id_24741]
purchase_history_24741_df[['order_id', 'category', 'subcategory']]
purchase_history_24741_df['category'].value_counts()

category
consumables    95
furniture      47
decor          30
software       28
electronics     7
Name: count, dtype: int64

Summary:

In part one, I counted the number of entries for each client to determine our top 5 clients. Client ID 33615 placed first with a total of 220 entries.

I digged deeper and calculated the total amount spent per client. Although Client 33615 had the most entries, Client ID 24741 spent a total of $82.27 million dollars. Client ID 33615 spent $8.38 million dollars.

After we deduct the cost to ship out the product and the cost to purchase the item, the company made a profit of $36.5 million dollars from Client ID 24741. 

I was curious to know if the item category type might have played a role in the total amount spent, but the results are inconclusive.